### DISCUSSION TITLE
Does ash use function pointers or link against the loader directly?

### DISCUSSION THREAD

**Discussion by Jerrody** [ash-rs/ash#683] [Q&A] [answered]
For example, in C/C++ headers if we call a PFN we skip the step when loader choose what driver to call.

Do I need to call a PFN or ash does all work?

**Comment by MarijnS95** [comment]
@Jerrody Are you referring whether a function is loaded generically through `vkGetInstanceProcAddr` _with or without_ `VkInstance`, either resulting in an ICD loader dispatch wrapper or a function specific for that instance?

https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetInstanceProcAddr.html#_description

Same for `vkGetDeviceProcAddr`: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetDeviceProcAddr.html#_description

**Comment by Jerrody** [comment]
@MarijnS95 

I'm not big professional I refer to this article where shown that calling of `vkCmdDraw` is an overhead, instead we call directly `PFN_vkCmdDraw` at this case we remove this overhead.

https://gpuopen.com/learn/reducing-vulkan-api-call-overhead/

So, my question is does `ash` call directly this `PFN` where we call `device. cmd_draw()`?

**Comment by MarijnS95** [comment, marked as answer]
@Jerrody `PFN_vkCmdDraw` it the _type of_ a function pointer. The loader exports a function with this signature as `vkCmdDraw`, to be linked against directly but with aforementioned dispatch table overhead. Ash does _not_ use those and instead loads it via `vkGetDeviceProcAddr(device, "vkCmdDraw")` directly, as written in that article:


https://github.com/ash-rs/ash/blob/29b935b1c176d6d53696beeaf09b7735ab4af5e7/ash/src/vk/features.rs#L3051-L3052

`_f` here is the loader function:

https://github.com/ash-rs/ash/blob/29b935b1c176d6d53696beeaf09b7735ab4af5e7/ash/src/device.rs#L22-L24

Which is used when creating a device:

https://github.com/ash-rs/ash/blob/29b935b1c176d6d53696beeaf09b7735ab4af5e7/ash/src/instance.rs#L341-L357

tl;dr: Ash uses optimized function pointers (specific to an instance/device) in almost all cases [^1].

[^1]: Some extensions combine device-level and instance-level functions: the latter cannot be loaded for a `device`, hence the entire extension is loaded via the `instance` and we take the dispatch table hit until our generator is capable of separating them out (https://github.com/ash-rs/ash/pull/556#discussion_r784237840).

**Comment by Jerrody** [comment]
@MarijnS95 

Thanks for the answer!
